glide-core/src/client/mod.rs

```diff
@@ -95,13 +95,89 @@ pub struct Client {
 }
 
 async fn run_with_timeout<T>(
-    timeout: Duration,
+    timeout: Option<Duration>,
     future: impl futures::Future<Output = RedisResult<T>> + Send,
 ) -> redis::RedisResult<T> {
-    tokio::time::timeout(timeout, future)
-        .await
-        .map_err(|_| io::Error::from(io::ErrorKind::TimedOut).into())
-        .and_then(|res| res)
+    match timeout {
+        Some(duration) => tokio::time::timeout(duration, future)
+            .await
+            .map_err(|_| io::Error::from(io::ErrorKind::TimedOut).into())
+            .and_then(|res| res),
+        None => future.await,
+    }
+}
+
+// Extension to the request timeout for blocking commands to ensure we won't return with timeout error before the server responded
+const BLOCKING_CMD_TIMEOUT_EXTENSION: f64 = 0.5; // seconds
+
+enum TimeUnit {
+    Milliseconds = 1000,
+    Seconds = 1,
+}
+
+// Enumeration representing different request timeout options.
+#[derive(Default, PartialEq, Debug)]
+enum RequestTimeoutOption {
+    // Indicates no timeout should be set for the request.
+    NoTimeout,
+    // Indicates the request timeout should be based on the client's configured timeout.
+    #[default]
+    ClientConfig,
+    // Indicates the request timeout should be based on the timeout specified in the blocking command.
+    BlockingCommand(Duration),
+}
+
+// Attempts to get the timeout duration from the command argument at `timeout_idx`.
+// If the argument can be parsed into a duration, it returns the duration in seconds with BlockingCmdTimeout.
+// If the timeout argument value is zero, NoTimeout will be returned. Otherwise, ClientConfigTimeout is returned.
+fn get_timeout_from_cmd_arg(
+    cmd: &Cmd,
+    timeout_idx: usize,
+    time_unit: TimeUnit,
+) -> RequestTimeoutOption {
+    cmd.arg_idx(timeout_idx)
+        .and_then(|timeout_bytes| {
+            std::str::from_utf8(timeout_bytes)
+                .ok()
+                .and_then(|timeout_str| {
+                    timeout_str.parse::<f64>().ok().map(|timeout| {
+                        let mut timeout_secs = timeout / ((time_unit as i32) as f64);
+                        if timeout_secs < 0.0 {
+                            // Timeout cannot be negative, return the client's configured request timeout
+                            return RequestTimeoutOption::ClientConfig;
+                        } else if timeout_secs == 0.0 {
+                            // `0` means we should set no timeout
+                            return RequestTimeoutOption::NoTimeout;
+                        } else {
+                            // Extend the request timeout to ensure we don't timeout before receiving a response from the server.
+                            timeout_secs += BLOCKING_CMD_TIMEOUT_EXTENSION;
+                        };
+                        RequestTimeoutOption::BlockingCommand(Duration::from_secs_f64(timeout_secs))
```

Comment:
We can avoid making timeout_secs mutable and also avoid the early returns here.
```suggestion
                        let timeout_secs = timeout / ((time_unit as i32) as f64);
                        if timeout_secs < 0.0 {
                            // Timeout cannot be negative, return the client's configured request timeout
                            RequestTimeoutOption::ClientConfig;
                        } else if timeout_secs == 0.0 {
                            // `0` means we should set no timeout
                            RequestTimeoutOption::NoTimeout;
                        } else {
                            // Extend the request timeout to ensure we don't timeout before receiving a response from the server.
                            RequestTimeoutOption::BlockingCommand(Duration::from_secs_f64(timeout_secs + BLOCKING_CMD_TIMEOUT_EXTENSION))
                        };
```